File: v2/220.py



# class Solution:
#     def containsNearbyAlmostDuplicate(self, nums, k, t):
#         """
#         :type nums: List[int]
#         :type k: int
#         :type t: int
#         :rtype: bool
#         """
#         # 6 star, 有左右两个节点，右节点从第二个到最后一个，左节点则是右节点往左k个之间的范围
#         # 超时了
#         length = len(nums)
#         right = 1
#         while right < length:
#             for left in range(max(0, right-k), right):
#                 if abs(nums[left] - nums[right]) <= t:
#                     return True
#             right += 1
#         return False



class Solution:
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        """
        :type nums: List[int]
        :type k: int
        :type t: int
        :rtype: bool
        """
        # 6 star, 有左右两个节点，右节点从第二个到最后一个，左节点则是右节点往左k个之间的范围
        if t==0 and len(nums)== len(set(nums)):
            return False
        for i in range(len(nums)):
            for j in range(i+1, i+k+1):
                if j>=len(nums):
                    break
                if abs(nums[i]-nums[j])<=t:
                    return True
        return False

File: go/220.go

package main

import (
	"math"
	//"fmt"
)

// 可以算是双指针法的变形，右指针从左到右遍历，左指针从右指针往前k个位置到右指针这段区间遍历，遍历时比较左右指针上的值的差是否小于t
func containsNearbyAlmostDuplicate(nums []int, k int, t int) bool {
	length := len(nums)
	for index:=1; index < length; index++ {
		left := index - k
		if left < 0 {left=0}
		for left < index {
			if math.Abs(float64(nums[left] - nums[index])) <= float64(t) {
				return true
			} else {
				left++
				}
		}
	}
	return false

}


//func main() {
//	nums := []int {1,2,3, 1}
//	ret := containsNearbyAlmostDuplicate(nums, 3, 0)
//	fmt.Println(ret)
//}